### PR TITLE
Replace CamelRuntimeProducer with RuntimeBeanBuildItem

### LINKED ISSUE #1672: Should the CamelRuntimeProducer be marked as unremovable?
A user reported having this in his logs:
```
2019-03-25 12:20:39,124 WARN  [io.qua.arc.run.ArcDeploymentTemplate] (main) Bean matching class io.quarkus.camel.core.runtime.CamelRuntimeProducer was marked as unused and removed during build.
Extensions can eliminate false positives using:
    - a custom UnremovableBeanBuildItem
    - AdditionalBeanBuildItem(false, beanClazz)
```

I think the `CamelRuntimeProducer` bean should probably be marked as unremovable.

### PR DESCRIPTION
**Pull request by mkouba**
- fix qualifier annotation values in RuntimeBeanProcessor
- add RuntimeBeanBuildItem.removable
- resolves #1672

### CODE REVIEW

File: core/deployment/src/main/java/io/quarkus/deployment/recording/BytecodeRecorderImpl.java [hunk 1]
```diff
@@ -255,7 +255,8 @@ public Object invoke(Object proxy, Method method, Object[] args) throws Throwabl
                         return "Runtime proxy of " + returnType + " with id " + key;
                     }
                     throw new RuntimeException(
-                            "You cannot invoke directly on an object returned from the bytecode recorded, you can only pass is back into the recorder as a parameter");
+                            "You cannot invoke " + method.getName()
+                                    + "() directly on an object returned from the bytecode recorded, you can only pass is back into the recorder as a parameter");
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/recording/BytecodeRecorderImpl.java]: They are preexisting but:
- I'm wondering if `from the bytecode recorded` should instead be `from the bytecode recorder`?
- s/pass is back/pass it back/

File: extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/RuntimeBeanBuildItem.java [hunk 2]
```diff
@@ -102,8 +113,18 @@ public Builder setRemovable(boolean removable) {
...
+        }
+
+        public Builder setRuntimeValue(RuntimeValue<?> runtimeValue) {
+            this.value = Objects.requireNonNull(runtimeValue);
+            return this;
+        }
+
         public RuntimeBeanBuildItem build() {
-            return new RuntimeBeanBuildItem(scope, type, supplier, qualifiers, removable);
+            return new RuntimeBeanBuildItem(scope, type, supplier, qualifiers, removable, value);
```
Comment (gsmet) [inline extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/RuntimeBeanBuildItem.java]: I think I would have put `supplier` together with `value` at the end of the constructor.

File: extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/RuntimeBeanProcessor.java [hunk 3]
```diff
@@ -48,7 +48,11 @@ public void write(String name, byte[] data) {
             //deterministic name
             //as we know the maps are sorted this will result in the same hash for the same bean
             String name = bean.type.replace(".", "_") + "_" + HashUtil.sha1(bean.qualifiers.toString());
-            map.put(name, bean.supplier);
+            if (bean.runtimeValue != null) {
+                map.put(name, template.createSupplier(bean.runtimeValue));
```
Comment (gsmet) [inline extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/RuntimeBeanProcessor.java]: Wondering why you don't create the supplier here directly?

File: extensions/camel/camel-core/deployment/src/main/java/io/quarkus/camel/core/deployment/CamelRuntimeBuildItem.java [hunk 4]
```diff
@@ -3,15 +3,17 @@
...
 
-    public CamelRuntimeBuildItem(CamelRuntime runtime) {
+    private final RuntimeValue<CamelRuntime> runtime;
+
+    public CamelRuntimeBuildItem(RuntimeValue<CamelRuntime> runtime) {
         this.runtime = runtime;
     }
 
-    public CamelRuntime getRuntime() {
+    public RuntimeValue<CamelRuntime> getRuntime() {
```
Comment (gsmet) [inline extensions/camel/camel-core/deployment/src/main/java/io/quarkus/camel/core/deployment/CamelRuntimeBuildItem.java]: Don't you think you should unwrap it here? I think the idea is really to get the `CamelRuntime`.

**Review by gsmet — COMMENTED** [review]
Added a few questions and comments inline.

File: core/deployment/src/main/java/io/quarkus/deployment/recording/BytecodeRecorderImpl.java (hunk 1 above)
Comment (mkouba) [reply core/deployment/src/main/java/io/quarkus/deployment/recording/BytecodeRecorderImpl.java]: Good catch.

File: extensions/camel/camel-core/deployment/src/main/java/io/quarkus/camel/core/deployment/CamelRuntimeBuildItem.java (hunk 4 above)
Comment (mkouba) [reply extensions/camel/camel-core/deployment/src/main/java/io/quarkus/camel/core/deployment/CamelRuntimeBuildItem.java]: Not really. The point is to provide a proxy that could passed to a template method.

File: extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/RuntimeBeanProcessor.java (hunk 3 above)
Comment (mkouba) [reply extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/RuntimeBeanProcessor.java]: The map is passed as a param to a template method. And we cannot record a lambda instance or an instance of annonymous class.

File: extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/RuntimeBeanBuildItem.java (hunk 2 above)
Comment (mkouba) [reply extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/RuntimeBeanBuildItem.java]: No one will use this constructor directly so I don't think it really matters.

**Review by stuartwdouglas — CHANGES_REQUESTED** [review]
We should probably be validating that only one of runtimeValue and supplier is set, otherwise looks good

**Comment by mkouba** [comment]
@stuartwdouglas Good point. Right now, the runtimeValue has precedence but it would be reasonable to fail if both are set.

**Review by stuartwdouglas — APPROVED** [review]

**Pull request merged**